File: unit-testing/python-good-coverage/src/calculator/calculator.py
"""
Simple Calculator class for basic arithmetic operations.

Taiga Task: TG-1 - Implement Calculator Class
User Story: As a user, I want to perform basic arithmetic operations
"""
import math


class Calculator:
    """Calculator class providing basic arithmetic operations."""

    def add(self, a: int, b: int) -> int:
        """
        Add two numbers.

        Args:
            a: First number
            b: Second number

        Returns:
            Sum of a and b
        """
        return a + b

    def subtract(self, a: int, b: int) -> int:
        """
        Subtract second number from first.

        Args:
            a: First number
            b: Second number

        Returns:
            Difference of a and b
        """
        return a - b

    def multiply(self, a: int, b: int) -> int:
        """
        Multiply two numbers.

        Args:
            a: First number
            b: Second number

        Returns:
            Product of a and b
        """
        return a * b

    def divide(self, a: int, b: int) -> float:
        """
        Divide first number by second.

        Args:
            a: Dividend
            b: Divisor

        Returns:
            Quotient of a divided by b

        Raises:
            ArithmeticError: If b is zero
        """
        if b == 0:
            raise ArithmeticError("Cannot divide by zero")
        return a / b

    def power(self, base: float, exponent: int) -> float:
        """
        Calculate the power of a number.

        Args:
            base: Base number
            exponent: Power to raise to

        Returns:
            Base raised to the power of exponent
        """
        return math.pow(base, exponent)

    def square_root(self, number: float) -> float:
        """
        Calculate square root of a number.

        Args:
            number: The number to find square root of

        Returns:
            Square root of the number

        Raises:
            ValueError: If number is negative
        """
        if number < 0:
            raise ValueError("Cannot calculate square root of negative number")
        return math.sqrt(number)

File: project-tools/jenkins-examples/Example_1/src/main/java/com/asu/ser516/Calculator.java
package com.asu.ser516;

/**
 * Simple Calculator class for basic arithmetic operations.
 * 
 * Taiga Task: TG-1 - Implement Calculator Class
 * User Story: As a user, I want to perform basic arithmetic operations
 */
public class Calculator {

    /**
     * Adds two numbers
     * @param a first number
     * @param b second number
     * @return sum of a and b
     */
    public int add(int a, int b) {
        return a + b;
    }

    /**
     * Subtracts second number from first
     * @param a first number
     * @param b second number
     * @return difference of a and b
     */
    public int subtract(int a, int b) {
        return a - b;
    }

    /**
     * Multiplies two numbers
     * @param a first number
     * @param b second number
     * @return product of a and b
     */
    public int multiply(int a, int b) {
        return a * b;
    }

    /**
     * Divides first number by second
     * @param a dividend
     * @param b divisor
     * @return quotient of a divided by b
     * @throws ArithmeticException if b is zero
     */
    public double divide(int a, int b) {
        if (b == 0) {
            throw new ArithmeticException("Cannot divide by zero");
        }
        return (double) a / b;
    }

    /**
     * Calculates the power of a number
     * @param base base number
     * @param exponent power to raise to
     * @return base raised to the power of exponent
     */
    public double power(double base, int exponent) {
        return Math.pow(base, exponent);
    }

    /**
     * Calculates square root of a number
     * @param number the number to find square root of
     * @return square root of the number
     * @throws IllegalArgumentException if number is negative
     */
    public double squareRoot(double number) {
        if (number < 0) {
            throw new IllegalArgumentException("Cannot calculate square root of negative number");
        }
        return Math.sqrt(number);
    }
}
